prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -966,6 +966,130 @@ void rxr_pkt_proc_received(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
 	}
 }
 
+/**
+ * @brief determine whether a received packet should be ignored
+ *
+ * There are two types of packets need to be ignored.
+ *
+ * First, packets from a removed peer should be ignored. Removed peer
+ * can happen when application called fi_av_remove to remove this address from
+ * address vector, which means application does not want to receive data
+ * from this peer any more. Thus packet from such peer should be ignored.
+ *
+ * Second, packets from a destroyed peer should be ignored. It is
+ * possible to receive packets from a peer that has already been
+ * destroyed on the remove instance, as demonstrated in the following
+ * workflow:
+ *
+ *     1. a peer was created on remote instance, with certain
+ *        GID + QPN.
+ *     2. the peer had communication with the endpoint, thus
+ *        endpoint has its address in AV.
+ *     3. the peer was destroyed. But before it is destroyed, it
+ *        sent some packets to the endpoint.
+ *     4. a new peer was created on remote instance, with the same GID + QPN.
+ *     5. the new peer sent some packets to the endpoint.
+ *     6. Upon receiving the packets from new peer, the endpoint removed
+ *        the address of old peer.
+ *     7. The packets from the old (destroyed) peer arrived.
+ *
+ * For the packets from destroyed peer, because they are from same GID + QPN,
+ * their pkt_entry->addr is same as new peer. By using connid in the header,
+ * we can identify these packets and mark them as "should be ignored"
+ *
+ * This function also insert address for REQ packets from a newly created peer
+ * (packets received on step 5 in the above workflow), and set
+ * pkt_entry->addr of these packets properly.
+ *
+ * @param[in]	ep		endpoint
+ * @param[in]	pkt_entry	received packet entry
+ * @return	a boolean value indicating whether this packet should be ignored
+ */
+static inline
+bool rxr_pkt_should_ignore_received(struct rxr_ep *ep,
+				    struct rxr_pkt_entry *pkt_entry)
+{
+	uint32_t *connid;
+	struct rdm_peer *peer;
+	struct efa_ep_addr *peer_addr; /* current peer address */
+	struct efa_ep_addr *sender_addr; /* peer address in packet header */
+
+	if (rxr_get_base_hdr(pkt_entry->pkt)->type >= RXR_REQ_PKT_BEGIN &&
+	    rxr_pkt_req_raw_addr(pkt_entry)) {
+		/* A REQ packet with raw address header can be the 1st packet
+		 * the endpoint received from a peer. Thus we need to be
+		 * careful here:
+		 *
+		 * the packet can be ignored only if its connid
+		 * matches previous qkey. In this case, we know for sure
+		 * this packet is from a destroyed peer.
+		 *
+		 * In all other case, we need to update AV with the new address
+		 * and process the packet.
+		 */
+		peer = rxr_ep_get_peer(ep, pkt_entry->addr);
+		peer_addr = rxr_peer_raw_addr(ep, pkt_entry->addr);
+
+		sender_addr = rxr_pkt_req_raw_addr(pkt_entry);
+		assert(sender_addr);
+		if (peer && sender_addr->qkey == peer->prev_qkey) {
```

Comment:
I am hesitant to do that because it will have bigger impact on performance, we will have to do a search of connid for every req packet we receive. I think I will ask internally what is the chance of that (three or more connid pending on the network) happenning.